File: deseq2/R/preprocessing.R
#' @title Preparar Datos para DESeq2
#' @description Carga, filtra y formatea la matriz de conteos y los metadatos.
#' @param gct_path Ruta al archivo de conteos en formato GCT (comprimido en .gz).
#'               - Formato esperado: Archivo de texto con genes en filas y muestras en columnas.
#'                 Las primeras 2 líneas son cabeceras GCT. La primera columna debe ser 'Name' (IDs de genes).
#' @param metadata_path Ruta al archivo de metadatos.
#'                - Formato esperado: Archivo de texto delimitado por tabulaciones.
#'                  La primera columna debe contener los IDs de las muestras (y ser el nombre de la columna).
#' @param col_grupo Nombre de la columna en los metadatos que contiene la información de los grupos a comparar.
#' @param nombre_grupo1 String con el nombre del primer grupo de comparación. Es el grupo de referencia (el control)
#' @param nombre_grupo2 String con el nombre del segundo grupo de comparación.
#' @return Una lista con dos elementos: 'matriz_conteos' (la matriz de conteos filtrada y ordenada)
#'         y 'metadatos' (el data frame de metadatos filtrado y ordenado).

preparar_datos <- function(gct_path, metadata_path, col_grupo, nombre_grupo1, nombre_grupo2) {
  # --- Carga y filtrado de metadatos ---

  # Cargar el archivo completo de metadatos. Se asume que la primera columna contiene los IDs de muestra.
  all_metadata <- utils::read.delim(metadata_path, row.names = 1)

  # Validar que la columna y los grupos especificados existan.
  if (!col_grupo %in% colnames(all_metadata)) {
    stop(sprintf("ERROR: La columna '%s' no existe en el archivo de metadatos.", col_grupo))
  }
  niveles_grupo <- unique(all_metadata[[col_grupo]])
  if (!nombre_grupo1 %in% niveles_grupo) {
    stop(sprintf("ERROR: El grupo '%s' no se encuentra en la columna '%s'.", nombre_grupo1, col_grupo))
  }
  if (!nombre_grupo2 %in% niveles_grupo) {
    stop(sprintf("ERROR: El grupo '%s' no se encuentra en la columna '%s'.", nombre_grupo2, col_grupo))
  }

  # Filtrar metadatos para conservar únicamente las muestras de los grupos de interés.
  metadata_filtrado <- all_metadata %>%
    dplyr::filter(.data[[col_grupo]] %in% c(nombre_grupo1, nombre_grupo2))

  cat(sprintf("INFO - Muestras encontradas: %d para '%s' y %d para '%s'.\n",
              sum(metadata_filtrado[[col_grupo]] == nombre_grupo1), nombre_grupo1,
              sum(metadata_filtrado[[col_grupo]] == nombre_grupo2), nombre_grupo2))


  # --- Carga y filtrado de la matriz de conteos ---

  # Cargar el archivo, detectando el tipo de archivo por su extensión
  if (endsWith(gct_path, ".gct")) {
    # Si es un archivo GCT, saltamos las 2 primeras líneas.
    counts_tibble <- readr::read_tsv(gct_path, skip = 2, show_col_types = FALSE)
    
    # Procesamos el GCT: eliminamos 'Description' y usamos 'Name' para los nombres de filas.
    count_matrix <- counts_tibble %>%
      dplyr::select(-Description) %>%
      tibble::column_to_rownames("Name") %>%
      as.matrix()
    
  } else if (endsWith(gct_path, ".tsv") | endsWith(gct_path, ".txt")) {
    # Si es un archivo TSV o TXT, lo leemos directamente sin saltar líneas.
    counts_tibble <- readr::read_tsv(gct_path, show_col_types = FALSE)
    
    # Procesamos el TSV: la primera columna se usa para los nombres de filas, no hay 'Description'.
    # Se detecta el nombre de la primera columna para usarla en column_to_rownames.
    first_col_name <- colnames(counts_tibble)[1]
    
    count_matrix <- counts_tibble %>%
      tibble::column_to_rownames(first_col_name) %>%
      as.matrix()
      
  } else {
    stop("El formato de archivo no es soportado. Por favor, use .gct, .tsv, o .txt.")
  }

  # Filtrar la matriz de conteos para quedarnos solo con las columnas (muestras)
  # que están presentes en nuestros metadatos filtrados.
  muestras_a_mantener <- rownames(metadata_filtrado)
  count_matrix_filtrada <- count_matrix[, intersect(muestras_a_mantener, colnames(count_matrix))]

  # Sincronizar el orden: Reordenar los metadatos para que el orden de sus filas (muestras)
  # coincida exactamente con el orden de las columnas (muestras) en la matriz de conteos.
  # Esto es un requisito estricto de DESeq2.
  metadata_final <- metadata_filtrado[colnames(count_matrix_filtrada), ]

  # Convierte la columna a factor
  metadata_final[[col_grupo]] <- gsub(":", "_", metadata_final[[col_grupo]])
  metadata_final[[col_grupo]] <- factor(metadata_final[[col_grupo]])
  # Establece explícitamente cuál es el grupo de referencia (el control)
  # 'grupo1' debería ser tu grupo control o base.
  metadata_final[[col_grupo]] <- relevel(metadata_final[[col_grupo]], ref = gsub(":", "_", grupo1))

  # Devolver los datos listos para el análisis
  return(list(matriz_conteos = count_matrix_filtrada, metadatos = metadata_final))
}

File: deseq2/deseq2_python/preprocessing.py
# preprocessing.py
import pandas as pd


def prepare_data(gct_path, metadata_path, group_col, group1, group2):
    """
    Carga, valida, filtra y sincroniza los datos de conteos y metadatos.

    Returns:
        tuple: Una tupla conteniendo el DataFrame de conteos filtrado y el 
               DataFrame de metadatos final y sincronizado.
    """
    # Carga y filtrado de metadatos
    try:
        all_metadata = pd.read_csv(metadata_path, sep='\t', index_col=0)
    except FileNotFoundError:
        raise FileNotFoundError(f"ERROR: No se encontró el archivo de metadatos en '{metadata_path}'")

    # Validar que la columna y los grupos especificados existan
    if group_col not in all_metadata.columns:
        raise ValueError(f"ERROR: La columna '{group_col}' no existe en el archivo de metadatos.")

    unique_groups = all_metadata[group_col].unique()
    if group1 not in unique_groups:
        raise ValueError(f"ERROR: El grupo '{group1}' no se encuentra en la columna '{group_col}'.")
    if group2 not in unique_groups:
        raise ValueError(f"ERROR: El grupo '{group2}' no se encuentra en la columna '{group_col}'.")

    # Filtrar metadatos para conservar únicamente las muestras de los grupos de interés
    metadata_filtered = all_metadata[all_metadata[group_col].isin([group1, group2])].copy()

    print(f"INFO - Muestras encontradas: {(metadata_filtered[group_col] == group1).sum()} para '{group1}' y "
          f"{(metadata_filtered[group_col] == group2).sum()} para '{group2}'.")

    # Carga y filtrado de la matriz de conteos
    matrix_type = None
    try:
        if gct_path.endswith("gct") or gct_path.endswith("gct.gz"):
            matrix_type = "GCT"
            counts_df = pd.read_csv(gct_path, sep='\t', skiprows=2, index_col=0)
        elif gct_path.endswith("csv") or gct_path.endswith("csv.gz"):
            matrix_type = "CSV"
            counts_df = pd.read_csv(gct_path, index_col=0)
        elif gct_path.endswith("tsv") or gct_path.endswith("tsv.gz"):
            matrix_type = "TSV"
            counts_df = pd.read_csv(gct_path, sep='\t', index_col=0)
        elif gct_path.endswith("txt") or gct_path.endswith("txt.gz"):
            matrix_type = "TSV"
            counts_df = pd.read_csv(gct_path, sep='\t', index_col=0)
        else:
            raise ValueError("ERROR: El archivo de conteos debe estar en formato .gct, .csv, .tsv o .txt (o sus versiones comprimidas).")
        # index_col=0 indica que la primera columna del archivo (índice 0) debe usarse como 
        # el índice de la tabla (DataFrame). En este contexto, esta columna contiene los 
        # nombres de los genes
    except FileNotFoundError:
        raise FileNotFoundError(f"ERROR: No se encontró el archivo GCT en '{gct_path}'")

    # Limpiar y asegurar que la matriz es numérica
    # Elimina la columna llamada "Description" del DataFrame. Esta columna contiene 
    # descripciones de los genes que no son necesarias para el análisis numérico
    if matrix_type == "GCT" and 'Description' in counts_df.columns:
        count_matrix = counts_df.drop(columns=['Description']).apply(pd.to_numeric)
    else:
        count_matrix = counts_df.apply(pd.to_numeric)

    # Sincronizar muestras entre la matriz de conteos y los metadatos
    # Encuentra los nombres de las muestras que están presentes en ambos DataFrames (en las
    # columnas de count_matrix y en el índice de metadata_filtered). Esto es crucial para 
    # asegurar que solo se analicen las muestras que tienen tanto datos de conteo como 
    # información descriptiva (grupos).
    common_samples = count_matrix.columns.intersection(metadata_filtered.index)

    count_matrix_filtered = count_matrix[common_samples]
    metadata_final = metadata_filtered.loc[common_samples]

    # Asegurar que el orden es idéntico (requisito de PyDESeq2)
    # Reordena las filas del DataFrame de metadatos (metadata_final) para que sigan 
    # exactamente el mismo orden que las columnas del DataFrame de conteos (count_matrix_filtered). 
    # Algunas herramientas de análisis, como PyDESeq2, requieren que este orden sea idéntico
    # para funcionar correctamente.
    metadata_final = metadata_final.reindex(count_matrix_filtered.columns)

    # Convertir la columna de grupo a categórica y establecer el nivel de referencia (control)
    metadata_final[group_col] = pd.Categorical(metadata_final[group_col], categories=[group1, group2])

    return count_matrix_filtered, metadata_final